File: legacy/node/src/deploy/platform-push.ts
import { execSync } from "node:child_process";
import type { DeployPlatform } from "../types.js";
import { platformDeployCommand } from "./platform-detector.js";

export interface PlatformPushResult {
  success: boolean;
  platform: DeployPlatform;
  message: string;
}

export function pushToPlatform(root: string, platform: DeployPlatform): PlatformPushResult {
  if (platform === "unknown") {
    return {
      success: false,
      platform,
      message: "Unknown platform — set platform in deploy.config.json or add vercel.json / railway.toml",
    };
  }

  const cmd = platformDeployCommand(platform);
  try {
    execSync(cmd, { cwd: root, stdio: "pipe", encoding: "utf8" });
    return { success: true, platform, message: `Deployed via: ${cmd}` };
  } catch (err) {
    const msg = err instanceof Error ? err.message : "Deploy command failed";
    const stderr =
      err && typeof err === "object" && "stderr" in err
        ? String((err as { stderr?: Buffer }).stderr ?? "")
        : "";
    return {
      success: false,
      platform,
      message: `${msg}${stderr ? `\n${stderr.slice(0, 500)}` : ""}`,
    };
  }
}

File: backend/apps/deploy/platform_push.py
"""Run platform CLI deploy commands in the project directory."""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path

from .platform import platform_deploy_command


def push_to_platform(project_root: Path, platform: str) -> dict:
    if platform == "unknown":
        return {
            "success": False,
            "platform": platform,
            "message": "Unknown platform — run scan or add vercel.json / railway.toml",
        }

    cmd = platform_deploy_command(platform)
    if platform == "railway" and not shutil.which("railway"):
        return {
            "success": False,
            "platform": platform,
            "message": (
                "Railway CLI not installed — push from GitHub instead: connect the SilverFox repo "
                "in Railway dashboard, or run: npm i -g @railway/cli"
            ),
        }
    try:
        subprocess.run(
            cmd,
            cwd=project_root,
            shell=True,
            check=True,
            capture_output=True,
            text=True,
            timeout=600,
        )
        return {"success": True, "platform": platform, "message": f"Deployed via: {cmd}"}
    except subprocess.CalledProcessError as exc:
        stderr = (exc.stderr or "")[:500]
        stderr_suffix = f"\n{stderr}" if stderr else ""
        return {
            "success": False,
            "platform": platform,
            "message": f"{exc}{stderr_suffix}",
        }
    except subprocess.TimeoutExpired:
        return {"success": False, "platform": platform, "message": "Deploy command timed out (10m)"}
    except OSError as exc:
        return {"success": False, "platform": platform, "message": str(exc)}
